### PR TITLE
Fix Boost handling with CMake

### PR DESCRIPTION
**Pull request by evpobr**
* Add Boost dependencies as imported libraries. This properly adds Boost
includes, libraries and also definitions to target
* Add forgotten Boost::date_time library

### CODE REVIEW

**Comment by arvidn** [comment]
`posix_time_duration` is necessary when building with deprecated functions enabled. Perhaps this dependency should be conditional too. But perhaps that would add too much complexity.

**Comment by ssiloti** [comment]
I think it's worth it to make the date_time dep condition on the deprecated functions option. Requiring libraries which aren't actually required is obnoxious if you're trying to do a space optimized build (e.g. for a consumer router).

File: CMakeLists.txt [hunk 1]
```diff
@@ -678,8 +678,12 @@ else()
 endif()
 
 # Boost
-find_public_dependency(Boost REQUIRED COMPONENTS system date_time)
-target_link_libraries(torrent-rasterbar PUBLIC Boost::system Boost::date_time)
+find_public_dependency(Boost REQUIRED COMPONENTS system)
+target_link_libraries(torrent-rasterbar PUBLIC Boost::system)
+if (TORRENT_NO_DEPRECATE)
```
Comment (arvidn) [inline CMakeLists.txt]: This condition looks inverted. Also, where does this definition come from?
A brief look at this file, I can't see an option to disable deprecated functions in cmake.

Comment (arvidn) [reply CMakeLists.txt]: oh, I see. It's this one `deprecated-functions`. I was thrown off by the conditional around that option. I don't think the comment (or logic) there is right. the tests *can* be built without deprecated functions

Comment (evpobr) [reply CMakeLists.txt]: Can you tell me what exactly i need to put in `if ()`? I see no useful options to start from.

Comment (evpobr) [reply CMakeLists.txt]: The name is really confusing. I understand `NO_DEPRECATE` as "do not deprecate, allow old ABI".

Probably good name is `NO_DEPRECATED`.

**Comment by arvidn** [comment]
@zeule does this look right to you?

**Comment by zeule** [comment]
> I don't use macros from `LibtorrentMacros`, they make me angry.

Could you explain why, please? With your choice the option identifier and the description string are duplicated for `option()` and `add_feature_info()` calls.

File: CMakeLists.txt [hunk 2]
```diff
@@ -13,6 +13,11 @@ set (SOVERSION "10")
 
 include(GNUInstallDirs)
 include(GeneratePkgConfig)
+include(CMakeDependentOption)
+
+# Boost
+find_public_dependency(Boost REQUIRED COMPONENTS system)
```
Comment (zeule) [inline CMakeLists.txt]: Why don't you modify https://github.com/arvidn/libtorrent/blob/master/CMakeLists.txt#L681?

Comment (zeule) [reply CMakeLists.txt]: Never mind, pardon.

File: CMakeLists.txt [hunk 3]
```diff
@@ -13,6 +13,11 @@ set (SOVERSION "10")
...
+find_public_dependency(Boost REQUIRED COMPONENTS date_time)
```
Comment (zeule) [inline CMakeLists.txt]: The `REQUIRED` keyword has to be removed.

Comment (zeule) [reply CMakeLists.txt]: `date_time` is not an unconditional public dependency, but it is only when `deprecated-functions` is set (according to your code below).

File: CMakeLists.txt [hunk 4]
```diff
@@ -677,11 +685,6 @@ else()
 	target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_DISABLE_DHT)
 endif()
 
-# Boost
-find_public_dependency(Boost REQUIRED COMPONENTS system)
-target_include_directories(torrent-rasterbar PUBLIC ${Boost_INCLUDE_DIRS})
```
Comment (zeule) [inline CMakeLists.txt]: This and the next line were here for a reason: every new boost release breaks imported targets in `FindBoost.cmake`. Please revert this change.

**Comment by zeule** [comment]
@evpobr: Take a look at the Travis build log. The build fails.

**Comment by evpobr** [comment]
> @evpobr: Take a look at the Travis build log. The build fails.

Failed to link Boost::system.

**Comment by zeule** [comment]
> Failed to link Boost::system.

target_link_libraries(torrent-rasterbar PUBLIC ${Boost_SYSTEM_LIBRARY})

**Comment by evpobr** [comment]
> > Failed to link Boost::system.
> 
> target_link_libraries(torrent-rasterbar PUBLIC ${Boost_SYSTEM_LIBRARY})

Actually i forgot to `add target_link_libraries(torrent-rasterbar PUBLIC Boost::system)`.

**Comment by evpobr** [comment]
`Boost::date_time` is linked as you can see as target. No problems.

**Comment by zeule** [comment]
> `Boost::system`

I've told you already why this is a bad idea.

**Comment by evpobr** [comment]
> > `Boost::system`
> 
> I've told you already why this is a bad idea.

Explain please. If you're right i will fix my code.

**Comment by zeule** [comment]
Please: every time boost version increases, cmake needs an update in FindBoost.cmake. Until the update imported targets do not exist for the new boost version. Now suppose your changes go public, and one day boost gets an update. Then the cmake build is broken until the next cmake version bump.

**Comment by zeule** [comment]
Example of the required cmake change: https://gitlab.kitware.com/cmake/cmake/commit/364fb69adc8ef0dfe59e5209735561d451c2ec9d

**Comment by evpobr** [comment]
> Please: every time boost version increases, cmake needs an update in FindBoost.cmake. Until the update imported targets do not exist for the new boost version. Now suppose your changes go public, and one day boost gets an update. Then the cmake build is broken until the next cmake version bump.

Is it? I thought `FindBoost` module parses `version.hpp`:

```cmake
# ------------------------------------------------------------------------
#  Extract version information from version.hpp
# ------------------------------------------------------------------------

# Set Boost_FOUND based only on header location and version.
# It will be updated below for component libraries.
if(Boost_INCLUDE_DIR)
  if(Boost_DEBUG)
    message(STATUS "[ ${CMAKE_CURRENT_LIST_FILE}:${CMAKE_CURRENT_LIST_LINE} ] "
                   "location of version.hpp: ${Boost_INCLUDE_DIR}/boost/version.hpp")
  endif()
...
```

**Comment by evpobr** [comment]
Well, fails again. What happens, GitHub problems? They activate private repos.

File: CMakeLists.txt [hunk 5]
```diff
@@ -536,6 +545,7 @@ target_compile_definitions(torrent-rasterbar
 target_link_libraries(torrent-rasterbar
 	PUBLIC
 		Threads::Threads
+		Boost::system
```
Comment (zeule) [inline CMakeLists.txt]: ${Boost_SYSTEM_LIBRARY}

File: CMakeLists.txt [hunk 6]
```diff
@@ -577,17 +587,20 @@ if (APPLE)
...
 # these options require existing target
 feature_option(dht "enable support for Mainline DHT" ON)
-if(NOT build_tests) # tests require deprecated symbols
-	target_optional_compile_definitions(torrent-rasterbar PUBLIC FEATURE NAME deprecated-functions DEFAULT ON
-		DESCRIPTION "enable deprecated functions for backwards compatibility" DISABLED TORRENT_NO_DEPRECATE)
-endif()
+cmake_dependent_option(deprecated-functions "enable deprecated functions for backwards compatibility" ON "Boost_DATE_TIME_FOUND" OFF)
+add_feature_info(deprecated-functions deprecated-functions "enables deprecated functions for backwards compatibility")
+target_compile_definitions(torrent-rasterbar PUBLIC $<$<NOT:$<BOOL:${deprecated-functions}>>:TORRENT_NO_DEPRECATE>)
+target_link_libraries(torrent-rasterbar PUBLIC $<$<BOOL:${deprecated-functions}>:Boost::date_time>)
```
Comment (zeule) [inline CMakeLists.txt]: ${Boost_DATE_TIME_LIBRARY}

**Comment by arvidn** [comment]
@evpobr what was the conclusion?

**Comment by evpobr** [comment]
No sense to develop this PR further with changes @zeule requested.
